src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs

```diff
@@ -0,0 +1,323 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.Collections
+{
+    /// <summary>
+    /// Provides temporary storage for a collection of elements. This type is optimized for handling of small
+    /// collections, particularly for cases where the collection will eventually be discarded or used to produce an
+    /// <see cref="ImmutableArray{T}"/>.
+    /// </summary>
+    /// <remarks>
+    /// This type stores small collections on the stack, with the ability to transition to dynamic storage if/when
+    /// larger number of elements are added.
+    /// </remarks>
+    /// <typeparam name="T">The type of elements stored in the collection.</typeparam>
+    [NonCopyable]
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct TemporaryArray<T> : IDisposable
+    {
+        /// <summary>
+        /// The number of elements the temporary can store inline. Storing more than this many elements requires the
+        /// array transition to dynamic storage.
+        /// </summary>
+        private const int InlineCapacity = 4;
+
+        /// <summary>
+        /// The first inline element.
+        /// </summary>
+        /// <remarks>
+        /// This field is only used when <see cref="_builder"/> is <see langword="null"/>. In other words, this type
+        /// stores elements inline <em>or</em> stores them in <see cref="_builder"/>, but does not use both approaches
+        /// at the same time.
+        /// </remarks>
+        private T _item0;
+
+#pragma warning disable IDE0044 // Add readonly modifier
+#pragma warning disable CS0649 // Field 'name' is never assigned to, and will always have its default value
```

Comment:
do we need this anymore, now that you directly access these by name?

Reply:
➡️ No. Will remove and leave it for the follow-up.